design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+
+![Shader Registry](resources/fig1.png)
+
+The GLES version of the Shader Registry can encapsulate the simple (but fairly weird) logic of shader source lookup that exists now.
+
+#### The way it is now
+
+The layers more or less pull their shaders out of a compressed block of text.  Clever, but difficult to access.
+
+#### Required Changes
+
+We'll need to make a Shader Registry class, of course, but that's fairly simple.  It should be thread safe, but you're mostly going to be using it on the main thread.
+
+The layers should pull their shaders out of the shader on demand.  This will give developers a chance to replace or add shaders before they're needed.
+
+### Rendering Passes
+
+_Addresses core functionality [#5](#core)._
+
+Modern graphics pipelines use more than one pass to create a visual representation.  It's actually pretty unusual to have a single rendering pass in your toolkit.  These are used for a dizzying array of purposes, including lightning effects, shadows, atmosphere and many others.  We (Wet Dog Weather) use separate rendering passes to draw data intermingled with map elements.
+
+Without getting too deep into specifics, early rendering passes allow the developer to use the power of the rasterizer for their own data.  Later rendering passes are typically used to decorate the map with effects.
+
+#### The way it is now:
+The [low level rendering logic](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/renderer/renderer_impl.cpp) in the toolkit seems to only support one explicit rendering pass.  
+
+However, there is the [RenderPass](https://github.com/maplibre/maplibre-gl-native/blob/main/src/mbgl/gl/render_pass.hpp) so some notion of this exists in the toolkit, but it's not clear how complete that is.  It may depend on ordering to work things out, rather than the explicit command buffer filling and fences we would use in a more modern approach.
+
+In any case, we need this filled out more explicitly for the render targets.
+
+#### Required changes:
+To flexibly implement multiple rendering passes we will need to:
+- Keep track of them by name and order
+- Implement Render Targets (see below)
+- Outputs from one pass as inputs to another
+- Optimize out passes without geometry
+- Expose all of this outside the toolkit
+
+#### Benefits
+Similar to what was discussed above:
+- Using the rasterizer to sort out your data ahead of time.  We do this a lot with weather data.
+- Reducing the size of a very costly shader.  The bigger your shader, the longer it takes.  If you have some fancy effect, it can probably be rendered at lower resolution and then laid over the top of the final scene.
+- Doing post-process effects on the map.  In games you might think of lens warping, wipes, and such.  For mapping, this tends to be more merging data layers or things like glow.
```

Comment:
I see those as "nice-to-have" instead of being essential for the modularization.
Though we need a good plan on how to manage even the resources of a single render pass.

The SDKs are quite verbose with Vulkan/Metal.

Reply:
As mentioned above there kind of are multiple rendering passes inside the toolkit, they're just not quite as explicit as they should be.  So think of this more as sorting out those issues and making that logic accessible outside the toolkit as well.